src/server/twemcache-rs/src/worker/single.rs

```diff
@@ -0,0 +1,288 @@
+            Eviction::None => Policy::None,
+            Eviction::Random => Policy::Random,
+            Eviction::Fifo => Policy::Fifo,
+            Eviction::Cte => Policy::Cte,
+            Eviction::Util => Policy::Util,
+            Eviction::Merge => Policy::Merge {
+                max: config.segcache().merge_max(),
+                merge: config.segcache().merge_target(),
+                compact: config.segcache().compact_target(),
+            },
+        };
+
+        let data = SegCache::builder()
+            .power(config.segcache().hash_power())
+            .overflow_factor(config.segcache().overflow_factor())
+            .heap_size(config.segcache().heap_size())
+            .segment_size(config.segcache().segment_size())
+            .eviction(eviction)
+            .hasher(CacheHasher::default())
+            .build();
+
+        Ok(Self {
+            config,
+            poll,
+            message_receiver,
+            message_sender,
+            session_receiver,
+            session_sender,
+            sessions,
+            data,
+        })
+    }
+
+    /// Run the `Worker` in a loop, handling new session events
+    pub fn run(&mut self) {
+        let mut events = Events::with_capacity(self.config.worker().nevent());
+        let timeout = Some(std::time::Duration::from_millis(
+            self.config.worker().timeout() as u64,
+        ));
+
+        #[cfg(feature = "heap_dump")]
+        let mut ops = 0;
+        #[cfg(feature = "heap_dump")]
+        let mut seq = 0;
+
+        loop {
+            increment_counter!(&Stat::WorkerEventLoop);
+
+            self.data.expire();
+
+            // get events with timeout
+            if self.poll.poll(&mut events, timeout).is_err() {
+                error!("Error polling");
+            }
+
+            increment_counter_by!(
+                &Stat::WorkerEventTotal,
+                events.iter().count().try_into().unwrap(),
+            );
+
+            // process all events
+            for event in events.iter() {
+                let token = event.token();
+
+                // event for existing session
+                trace!("got event for session: {}", token.0);
+
+                // handle error events first
+                if event.is_error() {
+                    increment_counter!(&Stat::WorkerEventError);
+                    self.handle_error(token);
+                }
+
+                // handle write events before read events to reduce write buffer
+                // growth if there is also a readable event
+                if event.is_writable() {
+                    increment_counter!(&Stat::WorkerEventWrite);
+                    self.do_write(token);
+                }
+
+                // read events are handled last
+                if event.is_readable() {
+                    increment_counter!(&Stat::WorkerEventRead);
+                    let _ = self.do_read(token);
+
+                    #[cfg(feature = "heap_dump")]
```

Comment:
What's this for?

Reply:
This should be documented - and probably is missing a corresponding feature flag in the Cargo.toml

This is to enable periodic dumping of the segment state to a format that allows us to generate visualizations of segment occupancy. This feature does require more thought, so I can either pull it out or leave it behind a feature flag.

Reply:
I ripped this out, it does feel a little hacky still - and I'd rather add a proper implementation in the future